### PR TITLE
Allow not validating target

### PR DESCRIPTION
**Pull request by PragTob**
As discussed in https://github.com/elixir-mint/mint/issues/453

I tried to follow the guidance of `case_sensitive_headers` so
that these options are treated somewhat similarly.

I tried to keep with the patterns of surrounding code, if you want anything changed/I missed something - happy to do so!

In a separate commit I also included a small nicety as I saw the definition of the mint user agent duplicated over the tests. Happy to roll that back though.

Some questions/remarks inline.

Thanks for all your work! :green_heart: 

Tasks left for me:
* [ ] Check if this actually fixes what I reported in #453 :)

IMG_20220619_184733

### CODE REVIEW

File: lib/mint/http1.ex [hunk 1]
```diff
@@ -123,6 +124,10 @@ defmodule Mint.HTTP1 do
     * `:case_sensitive_headers` - (boolean) if set to `true` the case of the supplied
        headers in requests will be preserved. The default is to lowercase the headers
        because HTTP/1.1 header names are case-insensitive. *Available since v1.6.0*.
+    * `:skip_target_validation` - (boolean) if set to `true` the target of a request
+       will not be validated. You might want this if you deal with non standard-
+       conform URIs but need to preserve them. The default is to validate the request
+       target. *Available since v1.?.?*
```
Comment (PragTob) [inline lib/mint/http1.ex:130]: Added it to keep on par with `case_sensititve_headers` but I'm not sure if this would be v 1.7.0 if it was released (technically it's a feature).

Also, should I add it to the Changelog then? :thinking:

File: lib/mint/http1/request.ex [hunk 2]
```diff
@@ -3,9 +3,9 @@ defmodule Mint.HTTP1.Request do
 
   import Mint.HTTP1.Parse
 
-  def encode(method, target, headers, body) do
+  def encode(method, target, headers, body, skip_target_validation \\ false) do
```
Comment (PragTob) [inline lib/mint/http1/request.ex:6]: this duplicates the default but I didn't:

* want it to be forced to be passed along
* break the existing interface (although technically it's `@moduledoc false`)

Also opted for a plain boolean, could also be turned into an `opts`/kwlist but seemed simpler so far.

File: test/mint/http1/conn_test.exs [hunk 3]
```diff
@@ -760,6 +760,38 @@ defmodule Mint.HTTP1Test do
 
                """)
     end
+
+    @invalid_target "%%"
+    test "targets are validated by default", %{port: port, server_ref: server_ref} do
+      assert {:ok, conn} = HTTP1.connect(:http, "localhost", port)
```
Comment (PragTob) [inline test/mint/http1/conn_test.exs:766]: may be overkill as I only added the option to skip, but I wanted the tests to show both cases and I wanted to test on the `conn` level to show that the option is correctly passed on

File: test/mint/http1/conn_test.exs [hunk 4]
```diff
@@ -548,7 +548,7 @@ defmodule Mint.HTTP1Test do
              request_string("""
              GET / HTTP/1.1
              host: localhost:#{port}
-             user-agent: mint/#{Mix.Project.config()[:version]}
+             user-agent: #{mint_user_agent()}
```
Comment (PragTob) [inline test/mint/http1/conn_test.exs:551]: This is just a small convenience change I made in a separate commit - hope it's not too much noise but I also didn't wanna do it in a separate PR. Happy to roll it back!

**Comment by PragTob** [comment]
Sorry for the force push, forgot one guiding comment and wanted to have it gone form history :sweat_smile:

**Comment by PragTob** [comment]
Works as expected on the `mint` level:

```
iex(1)> {:ok, conn} = Mint.HTTP.connect(:http, "httpbin.org", 80); {:ok, conn, request_ref} = Mint.HTTP.request(conn, "GET", "?na=%%boom%%", [], "")
** (MatchError) no match of right hand side value: {:error, %Mint.HTTP1{host: "httpbin.org", port: 80, request: nil, streaming_request: nil, socket: #Port<0.5>, transport: Mint.Core.Transport.TCP, mode: :active, scheme_as_string: "http", case_sensitive_headers: false, skip_target_validation: false, requests: {[], []}, state: :open, buffer: "", proxy_headers: [], private: %{}, log: false}, %Mint.HTTPError{reason: {:invalid_request_target, "?na=%%boom%%"}, module: Mint.HTTP1}}
    (stdlib 6.1) erl_eval.erl:652: :erl_eval.expr/6
    iex:1: (file)
iex(1)> {:ok, conn2} = Mint.HTTP.connect(:http, "httpbin.org", 80, skip_target_validation: true); {:ok, conn, request_ref} = Mint.HTTP.request(conn2, "GET", "?na=%%boom%%", [], "")
{:ok,
 %Mint.HTTP1{
   host: "httpbin.org",
   port: 80,
   request: %{
     status: nil,
     version: nil,
     state: :status,
     connection: [],
     body: nil,
     ref: #Reference<0.1432147674.813432838.138316>,
     data_buffer: [],
     headers_buffer: [],
     content_length: nil,
     method: "GET",
     transfer_encoding: []
   },
   streaming_request: nil,
   socket: #Port<0.6>,
   transport: Mint.Core.Transport.TCP,
   mode: :active,
   scheme_as_string: "http",
   case_sensitive_headers: false,
   skip_target_validation: true,
   requests: {[], []},
   state: :open,
   buffer: "",
   proxy_headers: [],
   private: %{},
   log: false
 },
 #Reference<0.1432147674.813432838.138316>}
iex(2)> flush()
{:tcp, #Port<0.6>,
 "HTTP/1.1 400 Bad Request\r\nServer: awselb/2.0\r\nDate: Fri, 08 Nov 2024 11:01:57 GMT\r\nContent-Type: text/html\r\nContent-Length: 122\r\nConnection: close\r\n\r\n<html>\r\n<head><title>400 Bad Request</title></head>\r\n<body>\r\n<center><h1>400 Bad Request</h1></center>\r\n</body>\r\n</html>\r\n"}
:ok
```

Can't figure out how to get it through Req/Finch (I thought I could get them to my Finch pool via `conn_opts` but that seems to not work, but that's not for here to solve :) )

File: lib/mint/http1.ex (hunk 1 above)
Comment (whatyouhide) [inline lib/mint/http1.ex:130]:
```suggestion
    * `:skip_target_validation` - (boolean) if set to `true` the target of a request
       will not be validated. You might want this if you deal with non standard-
       conforming URIs but need to preserve them. The default is to validate the request
       target. *Available since v1.7.0*.
```

File: test/mint/http1/conn_test.exs [hunk 5]
```diff
@@ -997,4 +1029,6 @@ defmodule Mint.HTTP1Test do
   defp stream_message_bytewise(<<>>, conn, responses) do
     {:ok, conn, responses}
   end
+
+  defp mint_user_agent, do: "mint/#{Mix.Project.config()[:version]}"
```
Comment (whatyouhide) [inline test/mint/http1/conn_test.exs:1033]: You can probably do this at compile time in a module attr even

**Review by whatyouhide — COMMENTED** [review]
Looking already great 💟

File: lib/mint/http1/request.ex (hunk 2 above)
Comment (PragTob) [reply lib/mint/http1/request.ex:6]: I'll try moving it over, the case sensitive headers option is also handled purely in `Mint.HTTP` - wasn't sure about the separation of concerns. Thanks for the feedback!

Comment (PragTob) [reply lib/mint/http1/request.ex]: Moved it over after a short coffee break :coffee: 

One thing that is slightly odd now is that target validation is in `Mint.HTTP1` while header validation is still in `Mint.HTTP1.Request`:

```elixir
  defp encode_headers(headers) do
    Enum.reduce(headers, "", fn {name, value}, acc ->
      validate_header_name!(name)
      validate_header_value!(name, value)
      [acc, name, ": ", value, "\r\n"]
    end)
  end
```

Might be worth to move. The reason I didn't yet is because it'd mean enumerating the headers another time and I'm not sure if that's a performance concern that matters in `mints` context (generally it shouldn't matter for "normal" amounts of headers but yeah).
Also it'd mean me changing more code that I don't know super well :)

Happy to change or keep depending on what you think @whatyouhide !

File: test/mint/http1/conn_test.exs (hunk 5 above)
Comment (PragTob) [reply test/mint/http1/conn_test.exs]: Works with `[:version]` :)

File: lib/mint/http1/request.ex (hunk 2 above)
Comment (whatyouhide) [reply lib/mint/http1/request.ex]: Let's keep it this way, don't worry about it for now.

**Comment by whatyouhide** [comment]
@PragTob do we need to do this for HTTP/2 too?

**Comment by PragTob** [comment]
@whatyouhide I'm unsure :) I'm guessing so but I didn't find any similar code with HTTP2 on a cursory look and I'm also not as familiar with HTTP2, let me see if I can find something.

**Comment by PragTob** [comment]
@whatyouhide as best as I can see there is no validation going on in the HTTP2 implementation.

`path` is taken here: https://github.com/elixir-mint/mint/blob/8516bcda190a644a5fe960b13869107ff76ee9b3/lib/mint/http2.ex#L510-L527

and only passed on to this function where it is just added to the header:

https://github.com/elixir-mint/mint/blob/8516bcda190a644a5fe960b13869107ff76ee9b3/lib/mint/http2.ex#L1354-L1370

there are some further validations going on, but as far as I can see none of them are for the path in particular/for a similar problem.

**Comment by whatyouhide** [comment]
Coolsies, then we're good 🙃
